packages/clay-color-picker/src/ClayColorPicker.tsx

```diff
@@ -0,0 +1,260 @@
+	selectColor: 'Select a color',
+	selectionIs: 'Color selection is {0}',
+};
+
+interface Props {
+	/**
+	 * Labels for the aria attributes
+	 */
+	ariaLabels?: {
+		selectColor: string;
+		selectionIs: string;
+	};
+
+	/**
+	 * List of color hex values
+	 */
+	colors?: string[];
+
+	/**
+	 * The label describing the collection of colors in the menu
+	 */
+	label?: string;
+
+	/**
+	 * The input attribute for name
+	 */
+	name?: string;
+
+	/**
+	 * Callback for when the list of colors change
+	 */
+	onColorsChange?: (val: string[]) => void;
+
+	/**
+	 * Callback for when the selected color changes
+	 */
+	onValueChange?: (val: string) => void;
+
+	/**
+	 * Determines if the hex input should render
+	 */
+	showHex?: boolean;
+
+	/**
+	 * Determines if the native color picker should be used
+	 */
+	useNative?: boolean;
+
+	/**
+	 * Value of the selected color hex
+	 */
+	value?: string;
+}
+
+const ColorPicker: React.SFC<Props> = ({
+	ariaLabels = DEFAULT_ARIA_LABELS,
+	colors = [],
+	label,
+	name,
+	onColorsChange = () => {},
+	onValueChange = () => {},
+	showHex = false,
+	useNative = false,
+	value = '#FFFFFF',
+}) => {
+	const containerRef = useRef<HTMLDivElement>(null);
+	const inputRef = useRef<HTMLInputElement>(null);
+	const valueInputRef = useRef<HTMLInputElement>(null);
+
+	const [active, setActive] = useState(false);
+	const [hexInputValue, setHexInput] = useHexInput(value.substring(1, 7));
+
+	useEffect(() => {
+		document.activeElement !== inputRef.current
+			? setHexInput(value.substring(1, 7))
+			: undefined;
+	}, [value]);
+
+	useEffect(() => {
+		const handleClick = (event: MouseEvent) => {
+			if (
+				event.target instanceof Element &&
+				containerRef.current &&
+				!containerRef.current.contains(event.target)
+			) {
+				setActive(false);
+			}
+		};
+
+		const handleEsc = (event: KeyboardEvent) =>
+			event.keyCode === ESC_KEY_CODE ? setActive(false) : undefined;
+
+		window.addEventListener('mousedown', handleClick);
+		window.addEventListener('keydown', handleEsc);
+
+		return () => {
+			window.removeEventListener('mousedown', handleClick);
+			window.removeEventListener('keydown', handleEsc);
+		};
+	}, []);
+
+	return (
+		<div className="clay-color-picker" ref={containerRef}>
+			{name && (
+				<input
+					name={name}
+					onChange={e =>
+						useNative ? onValueChange(e.target.value) : null
+					}
+					ref={valueInputRef}
+					style={{display: 'none'}}
+					type={useNative ? 'color' : 'text'}
+					value={value}
+				/>
+			)}
+
+			<div className="input-group">
+				<div
+					className={`input-group-item input-group-item-shrink${
+						showHex ? ' input-group-prepend' : ''
+					}`}
+				>
+					<span className="input-group-text input-group-text-secondary">
+						<div className="open-control open-checkbox">
+							<Splotch
+								aria-label={ariaLabels.selectColor}
+								className="btn btn-secondary"
+								onClick={() =>
+									useNative && valueInputRef.current
```

Comment:
hey @bryceosterhaus, I see a lot of the use of arrow functions and many logics being done directly in `onXXXX`... I personally prefer to call a function with a better description of what they are doing, so I can better understand the code, even though the Attempt here is to minimize the code, we can impact the maintainability and readability and maybe it did not have as much effect on production, I think declaring these functions before JSX is easy to understand what the component is doing.

Reply:
As far as reading the code goes, I don't really need a name here to add my understanding because its obvious that the function in `onClick={() => {}}` is a `handleClick` kind of function. The main issue, for me, is that anonymous functions are less helpful in stack traces. Of the different ways of naming the functions, what do you prefer?

```javascript
// Old school: may require a `.bind()` if you need to use `this`
onClick={function handleClick() {
}}

// Arrows: above...
const handleClick = () => { /* ... */ };
// then...
onClick={handleClick}
```

Reply:
Although the name of the property say much I do not agree that it says everything the question is that we have arrows above, can simplify the name of the functions and add descriptions if necessary to improve understanding, in my view I think add the whole logic above JSX, creates a mess and can make a salad, I prefer to add less logic over the JSX and separate a little. But I prefer arrow functions above 🙂

Reply:
I'd be hesitant to say we always need to put function handlers above and non-anonymous. A lot of times we might have something as simple as `onClick={() => setActiveIdHook(id)}` and I would hate to have to always move that to the top. Also we will likely tend to just call the function `handleXXX`. Even for the sake of stack traces, I think its helpful to have the name of the function, but also if its a simple enough function it probably isnt a use deal? For example, do we really think this [inline anonymous function](https://github.com/liferay/clay/pull/1706/files#diff-2919976302e4677a0f68885ac3357a9aR32) really needs to be moved to the top? Seems a bit overkill for me and will make readability a bit more difficult.

Now, trying to find a middle ground since I also understand your concern about too much logic in the jsx, what if we decided something along the lines of _"if its an implicit return, inlining the function is fine"_. Granted, that is very subjective and probably can't be automated with current eslint rules. Or we say if the function is more than 3 lines, move to the top?

Reply:
hey @bryceosterhaus I understand that we have cases that are not really necessary, especially when we have only one function being called. Actually we have to have the good sense of not being able to abuse to adds JSX, I do not know if the rule of going from 3 lines should be the right one to decide this, I think it goes from case to case, we get some cases we could be calling a function that does not make much sense being in the prop `onClick` for example but can also have a function that makes full sense is there.

Summarizing what I mean in the end is that separating the two environments in my view helps more to understand the code before arriving in JSX. But I'm not saying we should remove all the logics that could be there.

Reply:
The `> 3 lines` heuristic sounds pretty good to me, especially if we pair it with guidance to use your judgment on a case-by-case basis. I know we like having simple rules that we can apply uniformly everywhere, but sometimes having an element of flexibility is actually for the best.

Reply:
It seems good to me too, to go with that simple rule but tied to your case-by-case judgment, it would be interesting if we could add those decisions somewhere so that the next people will look at it. Be a JS style guide...